src/service/gitlab_service.ts

```diff
@@ -0,0 +1,128 @@
+import {BaseService} from './base_service'
+import { Gitlab } from 'gitlab'
+import * as GitUrlParse from 'git-url-parse'
+import {RepoMetadata} from './modal/repo_metadata'
+import {GitSource, SecretType} from "./modal/gitsource"
+import {RepoCheck} from "./modal/response_model/repo_check"
+import {Branch, BranchList} from "./modal/response_model/branch_list"
+import {RepoFileList} from "./modal/response_model/repo_file_list"
+import {ResponseLanguageList} from "./modal/response_model/language_list"
+
+export class GitlabService extends BaseService{
+  client: any
+  repodata: any
+  metadata: any
+  projectID: any
+  constructor(gitsource: GitSource){
+    super(gitsource)
+    this.repodata = this.getRepoMetadata()
+    var token = this.getAuthProvider()
+    this.client = new Gitlab({
+      host: this.repodata.host,
+      token: token
+    })
+  }
+
+  getRepoMetadata() {
+    if (this.repodata) {
+      return this.repodata
+    }
+    this.metadata = this.getMetadata()
+    let host = this.metadata.protocol + "://" + this.metadata.source
+    return new RepoMetadata(
+        this.metadata.owner,
+        host,
+        this.gitsource.ref,
+        this.metadata.name
+    )
+  }
+
+  getMetadata(){
+    if (this.metadata) {
+      return this.metadata
+    }
+    return GitUrlParse(this.gitsource.url)
+  }
+
+  getAuthProvider = (): any => {
+    switch(this.gitsource.secretType) {
+      case SecretType.PERSONAL_ACCESS_TOKEN || SecretType.OAUTH:
+        return this.gitsource.secretContent
+      default:
+        return null;
+    }
+  }
+
+  async getProjectId(): Promise<any> {
+    try {
+        if(this.projectID) {
+          return this.projectID;
+        }
+        this.metadata = this.getMetadata()
+        const resp = await this.client.Projects.search(this.metadata.name)
+        resp.forEach(p => {
+          if (p.path_with_namespace === this.metadata.full_name) {
+            this.projectID = p.id
+          }
+        });
+        return this.projectID;
+    }catch (e) {
+        throw e;
+    }
+  }
+
+  async isRepoReachable(): Promise<RepoCheck> {
+    this.metadata = this.getMetadata()
+    try {
+        const resp = await this.client.Projects.search(this.metadata.name)
```

Comment:
@pratikjagrut Do you think this would work for following scenario ? 

1. User is trying to search `devconsole-operator` repo.
2. Gitlab doesn't have `devconsole-operator` repo but has `devconsole-operator-test`.
3. Gitlab response would be non zero length array(search response would contain `devconsole-operator-test`), 
4. Expected is that array `resp.length` should always be zero always in case invalid repo right ?

you might need to think of such scenarios and change the code.

Reply:
@pratikjagrut I don't see any changes to fix this

Reply:
@pratikjagrut I see it thanks :)